cilium/CFP-12345-high-scale-ipcache.md

```diff
@@ -0,0 +1,108 @@
+
+## Motivation
+
+We currently add the IP addresses of all pods in the cluster in the ipcache map
+of each node. Some Cilium users however have clustermeshes with more than one
+million pods where that approach doesn't scale. To support such large
+clustermeshes, we need to avoid adding all remote pod IP addresses in the
+ipcache.
+
+## Goals
+
+* Scale clustermeshes to >1M pods.
+* Grow the ipcache sub-linearly with the number of pods in the cluster.
+* Support ingress policies as usual.
+
+## Non-Goals
+
+* Support FQDN policies for egress traffic to entities inside the cluster. That
+  will come in a follow-up CFP.
+* Support toEndpoints policies on egress.
+* Support advanced features such as the egress gateway or IPsec.
+
+## Proposal
+
+### Walkthrough
+
+Let's walk through a pod-to-remote-pod (pod1@node1 -> pod2@node2) connection
+with policies enforced on both egress and ingress:
+
+1. Pod1 sends a DNS request for pod2.default.pod.cluster.local.
+2. The DNS request is allowed via a toEndpoints rule. See the next section for
+   how this works.
+3. The proxy on node1 intercepts the DNS request and response and populates the
+   FQDN cache. See FQDN changes required here.
+4. A CIDR rule is added in the ipcache and policy maps to allow traffic to
+   pod2's IP address.
+5. Pod1 sends a SYN packet to pod2's IP.
+6. The packet is allowed by the CIDR rule.
+7. The packet is encapsulated with “pod1 -> pod2” as the outer IP header and
+   the tunnel ID is set to pod1's security identity. See changes required for
+   encapsulation here.
+8. Node2 receives the encapsulated packet, retrieves the source security
+   identity and decapsulates.
+9. The decapsulated packet is sent to pod2's ingress policy enforcement via
+   tail call and the ingress policies are enforced using the retrieved security
+   identity.
+
+### Changes to the IPcache
+
+The goal is to reduce the size of the ipcache and in particular for the ipcache
+to not grow linearly with the number of pods in the cluster.
+
+We will keep in the ipcache:
+- Special identities for entities (host, remote nodes, etc.).
+- Identities for well-known identities. This is mostly to get DNS pod
+  identities. Since we will use FQDN policies on egress, we will need to
+  resolve domain names via the DNS pods. We can't use the FQDN policies to
+  allow connections to the DNS pods obviously so we'll have to rely on
+  toEndpoints for that.
+- CIDR identities. This is needed to implement egress FQDN policies.
+
+### Pod-to-pod Encapsulation
+
+The encapsulation for pod-to-pod traffic is only required to carry the source
+identity over to the destination node. However, since Cilium doesn't know how
+to map pods to their node IP addresses (since we don't have the ipcache), we
+can't use our usual encapsulation with node IPs.
+
+Instead, for the inner and outer headers, we will use the same IP addresses,
+the pod IP addresses. Setting the outer destination IP address is easy, we can
```

Comment:
There isn't a real blocker to using the node IP address as the source IP address. One reason the pod IP address was preferred is that it avoids having an asymmetric connection on the physical network. If using the node IP address, the physical network may have issues correlating forward and reply packets, which may cause issues in some cases.

I've added a paragraph to mention the rationale here.

Reply:
Ah okay great. On the other hand I was wondering whether this should be yet another dedicated feature, since anyone with Pod IP routability on their network could potentially benefit from this, regardless of whether they're dealing with high-scale IPCache or not.

Reply:
> anyone with Pod IP routability on their network could potentially benefit from this, regardless of whether they're dealing with high-scale IPCache or not.

Benefit from what? The IP-in-IP encapsulation to carry the source identity?

Reply:
Yes, the benefit outlined a couple of paragraphs down:

> It would also have been possible to use the node IP address as the outer source
IP. That would however result in an assymetric connection on the physical
network. Forward packets would have IPs IP_NodeA -> IP_PodB and replies
IP_NodeB -> IP_PodA. That may cause issues with middleboxes and observability
tooling on the physical network.

This is an observability impact, which seems unrelated to the CFP goals of extensibility. Currently this CFP ties the two properties together, but they seem like entirely separate properties that could be independently toggled.